bokehjs/src/lib/models/plots/state_manager.ts

```diff
@@ -0,0 +1,76 @@
+import {Signal0} from "core/signaling"
+import type {PlotView} from "./plot_canvas"
+import type {RangeInfo} from "./range_manager"
+import {Selection} from "../selections/selection"
+import type {DataRenderer} from "../renderers/data_renderer"
+
+export type StateInfo = {
+  range?: RangeInfo
+  selection: Map<DataRenderer, Selection>
+  dimensions: {
+    width: number
+    height: number
+  }
+}
+
+export class StateManager {
+  constructor(readonly parent: PlotView, readonly initial_state: StateInfo) {}
+
+  readonly changed: Signal0<this["parent"]> = new Signal0(this.parent, "state_changed")
+
+  protected history: {type: string, state: StateInfo}[] = []
+  protected index: number = -1
+
+  protected _do_state_change(index: number): void {
+    const state = this.history[index] != null ? this.history[index].state : this.initial_state
+
+    if (state.range != null)
+      this.parent.update_range(state.range)
+
+    if (state.selection != null)
+      this.parent.update_selection(state.selection)
```

Comment:
Why not just pass the `state` to a generic `parent.update_state(state)` and let the parent sort it out? This way the state manager is more tightly coupled to knowing specific details about the parent and its relation to state updates.

Reply:
I'll assume this improvement is included with mentioned in https://github.com/bokeh/bokeh/pull/10537#discussion_r516553257